exp/runtime/internal/controllers/extensionconfig_controller.go

```diff
@@ -48,18 +54,101 @@ func (r *Reconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, opt
 	if err != nil {
 		return errors.Wrap(err, "failed setting up with a controller manager")
 	}
+	err = mgr.Add(&WarmupRunnable{
+		Client:        r.Client,
+		APIReader:     r.APIReader,
+		RuntimeClient: r.RuntimeClient,
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed adding WarmupRunnable to controller manager")
+	}
 	return nil
 }
 
-func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
-	extension := &runtimev1.ExtensionConfig{}
-	err := r.Client.Get(ctx, req.NamespacedName, extension)
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var errs []error
+	// Requeue events when the registry is not ready.
+	if !r.RuntimeClient.IsReady() {
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	extensionConfig := &runtimev1.ExtensionConfig{}
+	err := r.Client.Get(ctx, req.NamespacedName, extensionConfig)
 	if err != nil {
 		if apierrors.IsNotFound(err) {
-			return ctrl.Result{}, nil
+			// ExtensionConfig not found. Remove from registry.
+			return r.reconcileDelete(ctx, extensionConfig)
 		}
 		// Error reading the object - requeue the request.
 		return ctrl.Result{}, err
 	}
+
+	// Handle deletion reconciliation loop.
+	if !extensionConfig.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, extensionConfig)
+	}
+
+	// Make a copy of the ExtensionConfig to use as a base for patching.
+	original := extensionConfig.DeepCopy()
+
+	// discoverExtensionConfig will return a discovered ExtensionConfig with the appropriate conditions.
+	if extensionConfig, err = discoverExtensionConfig(ctx, r.RuntimeClient, extensionConfig); err != nil {
+		errs = append(errs, err)
+	}
+
+	// Always patch the ExtensionConfig as it may contain updates in conditions.
+	if err = patchExtensionConfig(ctx, r.Client, original, extensionConfig); err != nil {
+		errs = append(errs, err)
+	}
+
+	if len(errs) != 0 {
+		return ctrl.Result{}, kerrors.NewAggregate(errs)
+	}
+
+	// Register the extensionConfig if it was found and patched without error.
+	if err = r.RuntimeClient.Register(extensionConfig); err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to register ExtensionConfig %s/%s", extensionConfig.Namespace, extensionConfig.Name)
+	}
 	return ctrl.Result{}, nil
 }
+
+func patchExtensionConfig(ctx context.Context, client client.Client, original, modified *runtimev1.ExtensionConfig, options ...patch.Option) error {
+	patchHelper, err := patch.NewHelper(original, client)
```

Comment:
I first thought this err is not handled :). Can we please move l.121 directly below l.116?

And then l.118 directly before l.125